pkg/target/manager.go

```diff
@@ -0,0 +1,527 @@
+	ErrNoShootTargeted   = errors.New("no shoot targeted")
+)
+
+type Manager interface {
+	CurrentTarget() (Target, error)
+
+	TargetGarden(name string) error
+	TargetProject(ctx context.Context, name string) error
+	TargetSeed(ctx context.Context, name string) error
+	TargetShoot(ctx context.Context, name string) error
+
+	GardenClient(t Target) (client.Client, error)
+	ProjectClient(t Target) (client.Client, error)
+	SeedClient(t Target) (client.Client, error)
+	ShootNamespaceClient(t Target) (client.Client, error)
+	ShootClusterClient(t Target) (client.Client, error)
+}
+
+type managerImpl struct {
+	config          *config.Config
+	targetProvider  TargetProvider
+	clientProvider  ClientProvider
+	kubeconfigCache KubeconfigCache
+}
+
+var _ Manager = &managerImpl{}
+
+func NewManager(config *config.Config, targetProvider TargetProvider, clientProvider ClientProvider, kubeconfigCache KubeconfigCache) (Manager, error) {
+	return &managerImpl{
+		config:          config,
+		targetProvider:  targetProvider,
+		clientProvider:  clientProvider,
+		kubeconfigCache: kubeconfigCache,
+	}, nil
+}
+
+func (m *managerImpl) CurrentTarget() (Target, error) {
+	return m.targetProvider.Read()
+}
+
+func (m *managerImpl) TargetGarden(gardenName string) error {
+	for _, g := range m.config.Gardens {
+		if g.Name == gardenName {
+			return m.patchTarget(func(t *targetImpl) error {
+				t.Garden = gardenName
+				t.Project = ""
+				t.Seed = ""
+				t.Shoot = ""
+
+				return nil
+			})
+		}
+	}
+
+	return fmt.Errorf("garden %q is not defined in gardenctl configuration", gardenName)
+}
+
+func (m *managerImpl) TargetProject(ctx context.Context, projectName string) error {
+	return m.patchTarget(func(t *targetImpl) error {
+		if t.Garden == "" {
+			return ErrNoGardenTargeted
+		}
+
+		gardenClient, err := m.clientForGarden(t.Garden)
+		if err != nil {
+			return fmt.Errorf("could not create Kubernetes client for garden cluster: %w", err)
+		}
+
+		// validate that the project exists
+		project, err := m.resolveProjectName(ctx, gardenClient, projectName)
+		if err != nil {
+			return fmt.Errorf("failed to resolve project: %w", err)
+		}
+
+		// validate the project
+		if err := m.validateProject(ctx, project); err != nil {
+			return fmt.Errorf("invalid project: %w", err)
+		}
+
+		t.Seed = ""
+		t.Project = projectName
+		t.Shoot = ""
+
+		return nil
+	})
+}
+
+func (m *managerImpl) resolveProjectName(ctx context.Context, gardenClient client.Client, projectName string) (*gardencorev1beta1.Project, error) {
+	// validate that the project exists
+	project := &gardencorev1beta1.Project{}
+	key := types.NamespacedName{Name: projectName}
+	err := gardenClient.Get(ctx, key, project)
+
+	return project, err
+}
+
+func (m *managerImpl) validateProject(ctx context.Context, project *gardencorev1beta1.Project) error {
+	if project.Spec.Namespace == nil || *project.Spec.Namespace == "" {
+		return errors.New("project has not yet been fully created")
```

Comment:
```suggestion
		return errors.New("project does not have a corresponding namespace set. Most likely it has not yet been fully created")
```
so that the users knows what went wrong, in case they want to debug